{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, symlink the FASTQ files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting /work/rr151/TF_perturbSeq50genes_BL/generate_seqspec10x5pv2_gex.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile /work/rr151/TF_perturbSeq50genes_BL/generate_seqspec10x5pv2_gex.py\n",
    "#!/usr/bin/env python3\n",
    "import csv\n",
    "import yaml\n",
    "import gzip\n",
    "import os\n",
    "import argparse\n",
    "\n",
    "# Custom representers so PyYAML outputs tags like !Assay\n",
    "def tag_representer(dumper, data):\n",
    "    # data should be a dict with exactly one key: \"!Tag\"\n",
    "    tag = list(data.keys())[0]\n",
    "    value = data[tag]\n",
    "    return dumper.represent_mapping(tag, value)\n",
    "# Register for the tags you use\n",
    "yaml.add_representer(dict, tag_representer)\n",
    "\n",
    "def build_seqspec(sample_name, reads):\n",
    "    \"\"\"\n",
    "    Build seqspec YAML for one sample, given its paired read files.\n",
    "    \"\"\"\n",
    "    reads.sort(key=lambda r: r['illumina_read_type'])  # Ensure R1 then R2\n",
    "    # Build !Read entries\n",
    "    read_entries = []\n",
    "    for r in reads:\n",
    "        read_tag = \"!Read\"\n",
    "        read_entry = {\n",
    "            read_tag:  {\n",
    "                \"read_id\": r[\"accession\"],\n",
    "                \"name\": r[\"read_names\"].strip(\"[]\\\"\"),\n",
    "                \"modality\": \"rna\",\n",
    "                \"primer_id\": \"truseq_read1\" if r[\"illumina_read_type\"] == \"R1\" else \"truseq_read2\",\n",
    "                \"min_len\": 26 if r[\"illumina_read_type\"] == \"R1\" else 90,\n",
    "                \"max_len\": 26 if r[\"illumina_read_type\"] == \"R1\" else 90,\n",
    "                \"strand\": \"pos\" if r[\"illumina_read_type\"] == \"R1\" else \"neg\",\n",
    "                \"files\": [\n",
    "                    {\"!File\": {\n",
    "                        \"file_id\": r[\"accession\"],\n",
    "                        \"filename\": f\"https://api.data.igvf.org/sequence-files/{r['accession']}/@@download/{r['accession']}.fastq.gz\",\n",
    "                        \"filesize\": int(r[\"file_size\"]),\n",
    "                        \"filetype\": r[\"file_format\"],\n",
    "                        \"url\": f\"https://api.data.igvf.org/sequence-files/{r['accession']}/@@download/{r['accession']}.fastq.gz\",\n",
    "                        \"urltype\": \"https\",\n",
    "                        \"md5\": r[\"md5sum\"]\n",
    "                    }}\n",
    "                ]\n",
    "            }\n",
    "        \n",
    "        }\n",
    "        read_entries.append(read_entry)\n",
    "    index5 = reads[0]['index5']\n",
    "    index7 = reads[0]['index7']\n",
    "    seqspec = {\n",
    "        \"!Assay\": {\n",
    "            \"seqspec_version\": \"0.3.0\",\n",
    "            \"assay_id\": \"10XCRISPR5prime\",\n",
    "            \"name\": \"10x CRISPR and RNA assay 5'\",\n",
    "            \"doi\": \"https://doi.org/\",\n",
    "            \"date\": \"25 July 2025\",\n",
    "            \"description\": \"10x single-cell RNA-seq and CRISPR 5'\",\n",
    "            \"modalities\": [\"rna\"],\n",
    "            \"lib_struct\": \"https://teichlab.github.io/scg_lib_structs/methods_html/10xChromium5vdjfb.html\",\n",
    "            \"sequence_protocol\": \"Illumina NextSeq 2000 (EFO:0010963), Illumina NovaSeq X Plus (EFO:0022841)\",\n",
    "            \"sequence_kit\": reads[0][\"sequencing_kit\"],\n",
    "            \"library_protocol\": [\n",
    "                {\"!LibProtocol\": {\n",
    "                    \"protocol_id\": \"Chromium Next GEM Single Cell 5' HT Reagent Kits v2 (Dual Index) (10x Genomics)\",\n",
    "                    \"name\": \"5' Gene Expression (GEX) Library Construction\",\n",
    "                    \"modality\": \"rna\"\n",
    "                }}\n",
    "            ],\n",
    "            \"library_kit\": \"10XCRISPR5prime\",\n",
    "            \"sequence_spec\": read_entries,\n",
    "            \"library_spec\": [\n",
    "                {\"!Region\": {\n",
    "                    \"parent_id\": \"rna\",\n",
    "                    \"region_id\": \"rna\",\n",
    "                    \"region_type\": \"rna\",\n",
    "                    \"name\": \"rna\",\n",
    "                    \"sequence_type\": \"joined\",\n",
    "                    \"sequence\": \"AATGATACGGCGACCACCGAGATCTACACNNNNNNNNNNACATCTTTCCCTACACGACGCTCTTCCGATCTNNNNNNNNNNNNNNNNXXXXXXXXXXTTTCTTATATGGGXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXAGATCGGAAGAGCACACGTCTGAACTCCAGTCACNNNNNNNNNNATCTCGTATGCCGTCTTCTGCTTG\",\n",
    "                    \"min_len\": 135,\n",
    "                    \"max_len\": 269,\n",
    "                    \"onlist\": None,\n",
    "                    \"regions\": [ \n",
    "                         {\"!Region\": {\"region_id\": \"truseq_read1\", \"region_type\": \"truseq_read1\", \"name\": \"truseq_read1\", \"sequence_type\": \"fixed\", \"sequence\": \"ACACTCTTTCCCTACACGACGCTCTTCCGATCT\", \"min_len\": 33, \"max_len\": 33, \"onlist\": None, \"regions\": None, \"parent_id\": \"rna\"}},\n",
    "                         {\"!Region\": {\"region_id\": \"cell_bc\", \"region_type\": \"barcode\", \"name\": \"cell_bc\", \"sequence_type\": \"onlist\", \"sequence\": \"NNNNNNNNNNNNNNNN\", \"min_len\": 16, \"max_len\": 16, \"onlist\": {\"!Onlist\": {\"file_id\": \"IGVFFI9487JPEN\", \"filename\": \"https://api.data.igvf.org/tabular-files/IGVFFI9487JPEN/@@download/IGVFFI9487JPEN.tsv.gz\", \"filetype\": \"tsv\", \"filesize\": 2200000, \"url\": \"https://api.data.igvf.org/tabular-files/IGVFFI9487JPEN/@@download/IGVFFI9487JPEN.tsv.gz\", \"urltype\": \"https\", \"md5\": \"f62a276e262fdd85262a889d0f48556b\"}}, \"regions\": None, \"parent_id\": \"rna\"}},\n",
    "                         {\"!Region\": {\"region_id\": \"umi\", \"region_type\": \"umi\", \"name\": \"umi\", \"sequence_type\": \"random\", \"sequence\": \"XXXXXXXXXX\", \"min_len\": 10, \"max_len\": 10, \"onlist\": None, \"regions\": None, \"parent_id\": \"rna\"}},\n",
    "                         {\"!Region\": {\"region_id\": \"tso\", \"region_type\": \"linker\", \"name\": \"Template switch oligo\", \"sequence_type\": \"fixed\", \"sequence\": \"TTTCTTATATGGG\", \"min_len\": 13, \"max_len\": 13, \"onlist\": None, \"regions\": None, \"parent_id\": \"rna\"}},\n",
    "                         {\"!Region\": {\"region_id\": \"cdna\", \"region_type\": \"cdna\", \"name\": \"cdna\", \"sequence_type\": \"random\", \"sequence\": \"X\"*90, \"min_len\": 1, \"max_len\": 90, \"onlist\": None, \"regions\": None, \"parent_id\": \"rna\"}},\n",
    "                         {\"!Region\": {\"region_id\": \"truseq_read2\", \"region_type\": \"truseq_read2\", \"name\": \"truseq_read2\", \"sequence_type\": \"fixed\", \"sequence\": \"AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC\", \"min_len\": 34, \"max_len\": 34, \"onlist\": None, \"regions\": None, \"parent_id\": \"rna\"}},\n",
    "                         {\"!Region\": {\"region_id\": \"illumina_p5\", \"region_type\": \"illumina_p5\", \"name\": \"Illumina P5\", \"sequence_type\": \"fixed\", \"sequence\": \"AATGATACGGCGACCACCGAGATCTACAC\", \"min_len\": 29, \"max_len\": 29, \"onlist\": None, \"regions\": None, \"parent_id\": \"rna\"}},\n",
    "                         {\"!Region\": {\"region_id\": \"illumina_p7\", \"region_type\": \"illumina_p7\", \"name\": \"Illumina P7\", \"sequence_type\": \"fixed\", \"sequence\": \"ATCTCGTATGCCGTCTTCTGCTTG\", \"min_len\": 24, \"max_len\": 24, \"onlist\": None, \"regions\": None, \"parent_id\": \"rna\"}},\n",
    "                         {\"!Region\": {\"region_id\": \"index5\", \"region_type\": \"index5\", \"name\": \"index5\", \"sequence_type\": \"fixed\", \"sequence\": index5, \"min_len\": 10, \"max_len\": 10, \"onlist\": None, \"regions\": None, \"parent_id\": \"rna\"}},\n",
    "                         {\"!Region\": {\"region_id\": \"index7\", \"region_type\": \"index7\", \"name\": \"index7\", \"sequence_type\": \"fixed\", \"sequence\": index7, \"min_len\": 10, \"max_len\": 10, \"onlist\": None, \"regions\": None, \"parent_id\": \"rna\"}}\n",
    "                    ]\n",
    "                }}\n",
    "            ]\n",
    "        }\n",
    "    }\n",
    "    return seqspec\n",
    "def generate_seqspecs(input_csv, output_dir):\n",
    "    os.makedirs(output_dir, exist_ok=True)\n",
    "    samples = {}\n",
    "    with open(input_csv, newline='') as csvfile:\n",
    "        reader = csv.DictReader(csvfile)\n",
    "        for row in reader:\n",
    "            prefix = os.path.basename(row[\"submitted_file_name\"]).split('_R')[0]\n",
    "            samples.setdefault(prefix, []).append(row)\n",
    "    for sample, reads in samples.items():\n",
    "        seqspec = build_seqspec(sample, reads)\n",
    "        output_file = os.path.join(output_dir, f\"seqspec_{sample}.yaml.gz\")\n",
    "        with gzip.open(output_file, 'wt', encoding='utf-8') as gzfile:\n",
    "           yaml.dump(seqspec,gzfile,sort_keys=False)\n",
    "        \n",
    "        print(f\"Generated: {output_file}\")\n",
    "    print(f\"\\nDone! Seqspecs saved in: {output_dir}\")\n",
    "def main():\n",
    "    parser = argparse.ArgumentParser(description=\"Generate seqspec YAMLs (gzipped) from CSV\")\n",
    "    parser.add_argument(\"-i\", \"--input\", required=True, help=\"Input CSV file\")\n",
    "    parser.add_argument(\"-o\", \"--output\", default=\"seqspec_files\", help=\"Output directory\")\n",
    "    args = parser.parse_args()\n",
    "    generate_seqspecs(args.input, args.output)\n",
    "if __name__ == \"__main__\":\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting /work/rr151/TF_perturbSeq50genes_BL/generate_seqspec10x5pv2-sgRNA.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile /work/rr151/TF_perturbSeq50genes_BL/generate_seqspec10x5pv2-sgRNA.py\n",
    "#!/usr/bin/env python3\n",
    "import csv\n",
    "import yaml\n",
    "import gzip\n",
    "import os\n",
    "import argparse\n",
    "\n",
    "# Custom representers so PyYAML outputs tags like !Assay\n",
    "def tag_representer(dumper, data):\n",
    "    # data should be a dict with exactly one key: \"!Tag\"\n",
    "    tag = list(data.keys())[0]\n",
    "    value = data[tag]\n",
    "    return dumper.represent_mapping(tag, value)\n",
    "# Register for the tags you use\n",
    "yaml.add_representer(dict, tag_representer)\n",
    "\n",
    "def build_seqspec(sample_name, reads):\n",
    "    \"\"\"\n",
    "    Build seqspec YAML for one sample, given its paired read files.\n",
    "    \"\"\"\n",
    "    reads.sort(key=lambda r: r['illumina_read_type'])  # Ensure R1 then R2\n",
    "    # Build !Read entries\n",
    "    read_entries = []\n",
    "    for r in reads:\n",
    "        read_tag = \"!Read\"\n",
    "        read_entry = {\n",
    "            read_tag:  {\n",
    "                \"read_id\": r[\"accession\"],\n",
    "                \"name\": r[\"read_names\"].strip(\"[]\\\"\"),\n",
    "                \"modality\": \"crispr\",\n",
    "                \"primer_id\": \"truseq_read1\" if r[\"illumina_read_type\"] == \"R1\" else \"truseq_read2\",\n",
    "                \"min_len\": 26 if r[\"illumina_read_type\"] == \"R1\" else 90,\n",
    "                \"max_len\": 26 if r[\"illumina_read_type\"] == \"R1\" else 90,\n",
    "                \"strand\": \"pos\" if r[\"illumina_read_type\"] == \"R1\" else \"neg\",\n",
    "                \"files\": [\n",
    "                    {\"!File\": {\n",
    "                        \"file_id\": r[\"accession\"],\n",
    "                        \"filename\": f\"https://api.data.igvf.org/sequence-files/{r['accession']}/@@download/{r['accession']}.fastq.gz\",\n",
    "                        \"filesize\": int(r[\"file_size\"]),\n",
    "                        \"filetype\": r[\"file_format\"],\n",
    "                        \"url\": f\"https://api.data.igvf.org/sequence-files/{r['accession']}/@@download/{r['accession']}.fastq.gz\",\n",
    "                        \"urltype\": \"https\",\n",
    "                        \"md5\": r[\"md5sum\"]\n",
    "                    }}\n",
    "                ]\n",
    "            }\n",
    "        \n",
    "        }\n",
    "        read_entries.append(read_entry)\n",
    "    index5 = reads[0]['index5']\n",
    "    index7 = reads[0]['index7']\n",
    "    seqspec = {\n",
    "        \"!Assay\": {\n",
    "            \"seqspec_version\": \"0.3.0\",\n",
    "            \"assay_id\": \"10XCRISPR5prime\",\n",
    "            \"name\": \"10x CRISPR and RNA assay 5'\",\n",
    "            \"doi\": \"https://doi.org/\",\n",
    "            \"date\": \"25 July 2025\",\n",
    "            \"description\": \"10x single-cell RNA-seq and CRISPR 5'\",\n",
    "            \"modalities\": [\"crispr\"],\n",
    "            \"lib_struct\": \"https://teichlab.github.io/scg_lib_structs/methods_html/10xChromium5vdjfb.html\",\n",
    "            \"sequence_protocol\": \"Illumina NextSeq 2000 (EFO:0010963), Illumina NovaSeq X Plus (EFO:0022841)\",\n",
    "            \"sequence_kit\": reads[0][\"sequencing_kit\"],\n",
    "            \"library_protocol\": [\n",
    "                {\"!LibProtocol\": {\n",
    "                    \"protocol_id\": \"Chromium Next GEM Single Cell 5' HT Reagent Kits v2 (Dual Index) (10x Genomics)\",\n",
    "                    \"name\": \"5' Gene Expression (GEX) Library Construction\",\n",
    "                    \"modality\": \"crispr\"\n",
    "                }}\n",
    "            ],\n",
    "            \"library_kit\": \"10XCRISPR5prime\",\n",
    "            \"sequence_spec\": read_entries,\n",
    "            \"library_spec\": [\n",
    "                {\"!Region\": {\n",
    "                    \"parent_id\": \"crispr\",\n",
    "                    \"region_id\": \"crispr\",\n",
    "                    \"region_type\": \"crispr\",\n",
    "                    \"name\": \"crispr\",\n",
    "                    \"sequence_type\": \"joined\",\n",
    "                    \"sequence\": \"AATGATACGGCGACCACCGAGATCTACACNNNNNNNNNNACACTCTTTCCCTACACGACGCTCTTCCGATCTNNNNNNNNNNNNNNNNXXXXXXXXXXTTTCTTATATGGGGNNNNNNNNNNNNNNNNNNNNGTTTAAGAGCTAAGCTGGAAACAGCATAGCAAGTTTAAATAAGGCTAGTCCGTTATCAACTTGAGATCGGAAGAGCACACGTCTGAACTCCAGTCACNNNNNNNNNNATCTCGTATGCCGTCTTCTGCTTG\",\n",
    "                    \"min_len\": 135,\n",
    "                    \"max_len\": 269,\n",
    "                    \"onlist\": None,\n",
    "                    \"regions\": [\n",
    "                        {\"!Region\": {\"region_id\": \"truseq_read1\", \"region_type\": \"truseq_read1\", \"name\": \"truseq_read1\", \"sequence_type\": \"fixed\", \"sequence\": \"ACACTCTTTCCCTACACGACGCTCTTCCGATCT\", \"min_len\": 33, \"max_len\": 33, \"onlist\": None, \"regions\": None, \"parent_id\": \"crispr\"}},\n",
    "                        {\"!Region\": {\"region_id\": \"cell_bc\", \"region_type\": \"barcode\", \"name\": \"cell_bc\", \"sequence_type\": \"onlist\", \"sequence\": \"NNNNNNNNNNNNNNNN\", \"min_len\": 16, \"max_len\": 16, \"onlist\": {\"!Onlist\": {\"file_id\": \"IGVFFI9487JPEN\", \"filename\": \"https://api.data.igvf.org/tabular-files/IGVFFI9487JPEN/@@download/IGVFFI9487JPEN.tsv.gz\", \"filetype\": \"tsv\", \"filesize\": 2200000, \"url\": \"https://api.data.igvf.org/tabular-files/IGVFFI9487JPEN/@@download/IGVFFI9487JPEN.tsv.gz\", \"urltype\": \"https\", \"md5\": \"f62a276e262fdd85262a889d0f48556b\"}}, \"regions\": None, \"parent_id\": \"crispr\"}},\n",
    "                        {\"!Region\": {\"region_id\": \"umi\", \"region_type\": \"umi\", \"name\": \"umi\", \"sequence_type\": \"random\", \"sequence\": \"XXXXXXXXXX\", \"min_len\": 10, \"max_len\": 10, \"onlist\": None, \"regions\": None, \"parent_id\": \"crispr\"}},\n",
    "                        {\"!Region\": {\"region_id\": \"tso\", \"region_type\": \"linker\", \"name\": \"Template switch oligo\", \"sequence_type\": \"fixed\", \"sequence\": \"TTTCTTATATGGG\", \"min_len\": 13, \"max_len\": 13, \"onlist\": None, \"regions\": None, \"parent_id\": \"crispr\"}},\n",
    "                        {\"!Region\": {\"region_id\": \"sgrna_target\", \"region_type\": \"sgrna_target\", \"name\": \"sgrna_target\", \"sequence_type\": \"onlist\", \"sequence\": \"NNNNNNNNNNNNNNNNNNNN\", \"min_len\": 20, \"max_len\": 20, \"onlist\": {\"!Onlist\": {\"file_id\": \"IGVFFI6638COVL\", \"filename\": \"https://api.data.igvf.org/reference-files/IGVFFI6638COVL/@@download/IGVFFI6638COVL.tsv.gz\", \"filetype\": \"tsv\", \"filesize\": 2200000, \"url\": \"https://api.data.igvf.org/reference-files/IGVFFI6638COVL/@@download/IGVFFI6638COVL.tsv.gz\", \"urltype\": \"https\", \"md5\": \"7e6d956b7bf03c4cc6d16ae5fed9661b\"}}, \"regions\": None, \"parent_id\": \"crispr\"}},\n",
    "                        {\"!Region\": {\"region_id\": \"sgrna_scaffold\", \"region_type\": \"linker\", \"name\": \"sgrna_scaffold\", \"sequence_type\": \"fixed\", \"sequence\": \"GTTTAAGAGCTAAGCTGGAAACAGCATAGCAAGTTTA\", \"min_len\": 37, \"max_len\": 37, \"onlist\": None, \"regions\": None, \"parent_id\": \"crispr\"}},\n",
    "                        {\"!Region\": {\"region_id\": \"feature_primer\", \"region_type\": \"custom_primer\", \"name\": \"feature_primer\", \"sequence_type\": \"fixed\", \"sequence\": \"AATAAGGCTAGTCCGTTATCAACTTG\", \"min_len\": 26, \"max_len\": 26, \"onlist\": None, \"regions\": None, \"parent_id\": \"crispr\"}},\n",
    "                        {\"!Region\": {\"region_id\": \"truseq_read2\", \"region_type\": \"truseq_read2\", \"name\": \"truseq_read2\", \"sequence_type\": \"fixed\", \"sequence\": \"AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC\", \"min_len\": 34, \"max_len\": 34, \"onlist\": None, \"regions\": None, \"parent_id\": \"crispr\"}},\n",
    "                        {\"!Region\": {\"region_id\": \"illumina_p5\", \"region_type\": \"illumina_p5\", \"name\": \"Illumina P5\", \"sequence_type\": \"fixed\", \"sequence\": \"AATGATACGGCGACCACCGAGATCTACAC\", \"min_len\": 29, \"max_len\": 29, \"onlist\": None, \"regions\": None, \"parent_id\": \"crispr\"}},\n",
    "                        {\"!Region\": {\"region_id\": \"illumina_p7\", \"region_type\": \"illumina_p7\", \"name\": \"Illumina P7\", \"sequence_type\": \"fixed\", \"sequence\": \"ATCTCGTATGCCGTCTTCTGCTTG\", \"min_len\": 24, \"max_len\": 24, \"onlist\": None, \"regions\": None, \"parent_id\": \"crispr\"}},\n",
    "                        {\"!Region\": {\"region_id\": \"index5\", \"region_type\": \"index5\", \"name\": \"index5\", \"sequence_type\": \"fixed\", \"sequence\": index5, \"min_len\": 10, \"max_len\": 10, \"onlist\": None, \"regions\": None, \"parent_id\": \"crispr\"}},\n",
    "                        {\"!Region\": {\"region_id\": \"index7\", \"region_type\": \"index7\", \"name\": \"index7\", \"sequence_type\": \"fixed\", \"sequence\": index7, \"min_len\": 10, \"max_len\": 10, \"onlist\": None, \"regions\": None, \"parent_id\": \"crispr\"}}\n",
    "                    ]\n",
    "                }}\n",
    "            ]\n",
    "        }\n",
    "    }\n",
    "    return seqspec\n",
    "def generate_seqspecs(input_csv, output_dir):\n",
    "    os.makedirs(output_dir, exist_ok=True)\n",
    "    samples = {}\n",
    "    with open(input_csv, newline='') as csvfile:\n",
    "        reader = csv.DictReader(csvfile)\n",
    "        for row in reader:\n",
    "            prefix = os.path.basename(row[\"submitted_file_name\"]).split('_R')[0]\n",
    "            samples.setdefault(prefix, []).append(row)\n",
    "    for sample, reads in samples.items():\n",
    "        seqspec = build_seqspec(sample, reads)\n",
    "        output_file = os.path.join(output_dir, f\"seqspec_{sample}.yaml.gz\")\n",
    "        with gzip.open(output_file, 'wt', encoding='utf-8') as gzfile:\n",
    "           yaml.dump(seqspec,gzfile,sort_keys=False)\n",
    "        \n",
    "        print(f\"Generated: {output_file}\")\n",
    "    print(f\"\\nDone! Seqspecs saved in: {output_dir}\")\n",
    "def main():\n",
    "    parser = argparse.ArgumentParser(description=\"Generate seqspec YAMLs (gzipped) from CSV\")\n",
    "    parser.add_argument(\"-i\", \"--input\", required=True, help=\"Input CSV file\")\n",
    "    parser.add_argument(\"-o\", \"--output\", default=\"seqspec_files\", help=\"Output directory\")\n",
    "    args = parser.parse_args()\n",
    "    generate_seqspecs(args.input, args.output)\n",
    "if __name__ == \"__main__\":\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting /work/rr151/TF_perturbSeq50genes_BL/generate_seqspec10x5pv3_gex.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile /work/rr151/TF_perturbSeq50genes_BL/generate_seqspec10x5pv3_gex.py\n",
    "#!/usr/bin/env python3\n",
    "import csv\n",
    "import yaml\n",
    "import gzip\n",
    "import os\n",
    "import argparse\n",
    "\n",
    "# Custom representers so PyYAML outputs tags like !Assay\n",
    "def tag_representer(dumper, data):\n",
    "    # data should be a dict with exactly one key: \"!Tag\"\n",
    "    tag = list(data.keys())[0]\n",
    "    value = data[tag]\n",
    "    return dumper.represent_mapping(tag, value)\n",
    "# Register for the tags you use\n",
    "yaml.add_representer(dict, tag_representer)\n",
    "\n",
    "def build_seqspec(sample_name, reads):\n",
    "    \"\"\"\n",
    "    Build seqspec YAML for one sample, given its paired read files.\n",
    "    \"\"\"\n",
    "    reads.sort(key=lambda r: r['illumina_read_type'])  # Ensure R1 then R2\n",
    "    # Build !Read entries\n",
    "    read_entries = []\n",
    "    for r in reads:\n",
    "        read_tag = \"!Read\"\n",
    "        read_entry = {\n",
    "            read_tag:  {\n",
    "                \"read_id\": r[\"accession\"],\n",
    "                \"name\": r[\"read_names\"].strip(\"[]\\\"\"),\n",
    "                \"modality\": \"rna\",\n",
    "                \"primer_id\": \"truseq_read1\" if r[\"illumina_read_type\"] == \"R1\" else \"truseq_read2\",\n",
    "                \"min_len\": 26 if r[\"illumina_read_type\"] == \"R1\" else 90,\n",
    "                \"max_len\": 26 if r[\"illumina_read_type\"] == \"R1\" else 90,\n",
    "                \"strand\": \"pos\" if r[\"illumina_read_type\"] == \"R1\" else \"neg\",\n",
    "                \"files\": [\n",
    "                    {\"!File\": {\n",
    "                        \"file_id\": r[\"accession\"],\n",
    "                        \"filename\": f\"https://api.data.igvf.org/sequence-files/{r['accession']}/@@download/{r['accession']}.fastq.gz\",\n",
    "                        \"filesize\": int(r[\"file_size\"]),\n",
    "                        \"filetype\": r[\"file_format\"],\n",
    "                        \"url\": f\"https://api.data.igvf.org/sequence-files/{r['accession']}/@@download/{r['accession']}.fastq.gz\",\n",
    "                        \"urltype\": \"https\",\n",
    "                        \"md5\": r[\"md5sum\"]\n",
    "                    }}\n",
    "                ]\n",
    "            }\n",
    "        \n",
    "        }\n",
    "        read_entries.append(read_entry)\n",
    "    index5 = reads[0]['index5']\n",
    "    index7 = reads[0]['index7']\n",
    "    seqspec = {\n",
    "        \"!Assay\": {\n",
    "            \"seqspec_version\": \"0.3.0\",\n",
    "            \"assay_id\": \"10XCRISPR5prime\",\n",
    "            \"name\": \"10x CRISPR and RNA assay 5'\",\n",
    "            \"doi\": \"https://doi.org/\",\n",
    "            \"date\": \"25 July 2025\",\n",
    "            \"description\": \"10x single-cell RNA-seq and CRISPR 5'\",\n",
    "            \"modalities\": [\"rna\"],\n",
    "            \"lib_struct\": \"https://teichlab.github.io/scg_lib_structs/methods_html/10xChromium5vdjfb.html\",\n",
    "            \"sequence_protocol\": \"Illumina NextSeq 2000 (EFO:0010963), Illumina NovaSeq X Plus (EFO:0022841)\",\n",
    "            \"sequence_kit\": reads[0][\"sequencing_kit\"],\n",
    "            \"library_protocol\": [\n",
    "                {\"!LibProtocol\": {\n",
    "                    \"protocol_id\": \"Chromium Next GEM Single Cell 5' HT Reagent Kits v2 (Dual Index) (10x Genomics)\",\n",
    "                    \"name\": \"5' Gene Expression (GEX) Library Construction\",\n",
    "                    \"modality\": \"rna\"\n",
    "                }}\n",
    "            ],\n",
    "            \"library_kit\": \"10XCRISPR5prime\",\n",
    "            \"sequence_spec\": read_entries,\n",
    "            \"library_spec\": [\n",
    "                {\"!Region\": {\n",
    "                    \"parent_id\": \"rna\",\n",
    "                    \"region_id\": \"rna\",\n",
    "                    \"region_type\": \"rna\",\n",
    "                    \"name\": \"rna\",\n",
    "                    \"sequence_type\": \"joined\",\n",
    "                    \"sequence\": \"AATGATACGGCGACCACCGAGATCTACACNNNNNNNNNNACATCTTTCCCTACACGACGCTCTTCCGATCTNNNNNNNNNNNNNNNNXXXXXXXXXXXXTTTCTTATATGGGXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXAGATCGGAAGAGCACACGTCTGAACTCCAGTCACNNNNNNNNNNATCTCGTATGCCGTCTTCTGCTTG\",\n",
    "                    \"min_len\": 135,\n",
    "                    \"max_len\": 275,\n",
    "                    \"onlist\": None,\n",
    "                    \"regions\": [ \n",
    "                         {\"!Region\": {\"region_id\": \"truseq_read1\", \"region_type\": \"truseq_read1\", \"name\": \"truseq_read1\", \"sequence_type\": \"fixed\", \"sequence\": \"ACACTCTTTCCCTACACGACGCTCTTCCGATCT\", \"min_len\": 33, \"max_len\": 33, \"onlist\": None, \"regions\": None, \"parent_id\": \"rna\"}},\n",
    "                         {\"!Region\": {\"region_id\": \"cell_bc\", \"region_type\": \"barcode\", \"name\": \"cell_bc\", \"sequence_type\": \"onlist\", \"sequence\": \"NNNNNNNNNNNNNNNN\", \"min_len\": 16, \"max_len\": 16, \"onlist\": {\"!Onlist\": {\"file_id\": \"IGVFFI1697XAXI\", \"filename\": \"https://api.data.igvf.org/tabular-files/IGVFFI1697XAXI/@@download/IGVFFI1697XAXI.tsv.gz\", \"filetype\": \"tsv\", \"filesize\": 6900000, \"url\": \"https://api.data.igvf.org/tabular-files/IGVFFI1697XAXI/@@download/IGVFFI1697XAXI.tsv.gz\", \"urltype\": \"https\", \"md5\": \"5033d08c9b8d5080d2e5f65a3896b53b\"}}, \"regions\": None, \"parent_id\": \"rna\"}},\n",
    "                         {\"!Region\": {\"region_id\": \"umi\", \"region_type\": \"umi\", \"name\": \"umi\", \"sequence_type\": \"random\", \"sequence\": \"XXXXXXXXXXXX\", \"min_len\": 12, \"max_len\": 12, \"onlist\": None, \"regions\": None, \"parent_id\": \"rna\"}},\n",
    "                         {\"!Region\": {\"region_id\": \"tso\", \"region_type\": \"linker\", \"name\": \"Template switch oligo\", \"sequence_type\": \"fixed\", \"sequence\": \"TTTCTTATATGGG\", \"min_len\": 13, \"max_len\": 13, \"onlist\": None, \"regions\": None, \"parent_id\": \"rna\"}},\n",
    "                         {\"!Region\": {\"region_id\": \"cdna\", \"region_type\": \"cdna\", \"name\": \"cdna\", \"sequence_type\": \"random\", \"sequence\": \"X\"*90, \"min_len\": 1, \"max_len\": 90, \"onlist\": None, \"regions\": None, \"parent_id\": \"rna\"}},\n",
    "                         {\"!Region\": {\"region_id\": \"truseq_read2\", \"region_type\": \"truseq_read2\", \"name\": \"truseq_read2\", \"sequence_type\": \"fixed\", \"sequence\": \"AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC\", \"min_len\": 34, \"max_len\": 34, \"onlist\": None, \"regions\": None, \"parent_id\": \"rna\"}},\n",
    "                         {\"!Region\": {\"region_id\": \"illumina_p5\", \"region_type\": \"illumina_p5\", \"name\": \"Illumina P5\", \"sequence_type\": \"fixed\", \"sequence\": \"AATGATACGGCGACCACCGAGATCTACAC\", \"min_len\": 29, \"max_len\": 29, \"onlist\": None, \"regions\": None, \"parent_id\": \"rna\"}},\n",
    "                         {\"!Region\": {\"region_id\": \"illumina_p7\", \"region_type\": \"illumina_p7\", \"name\": \"Illumina P7\", \"sequence_type\": \"fixed\", \"sequence\": \"ATCTCGTATGCCGTCTTCTGCTTG\", \"min_len\": 24, \"max_len\": 24, \"onlist\": None, \"regions\": None, \"parent_id\": \"rna\"}},\n",
    "                         {\"!Region\": {\"region_id\": \"index5\", \"region_type\": \"index5\", \"name\": \"index5\", \"sequence_type\": \"fixed\", \"sequence\": index5, \"min_len\": 10, \"max_len\": 10, \"onlist\": None, \"regions\": None, \"parent_id\": \"rna\"}},\n",
    "                         {\"!Region\": {\"region_id\": \"index7\", \"region_type\": \"index7\", \"name\": \"index7\", \"sequence_type\": \"fixed\", \"sequence\": index7, \"min_len\": 10, \"max_len\": 10, \"onlist\": None, \"regions\": None, \"parent_id\": \"rna\"}}\n",
    "                    ]\n",
    "                }}\n",
    "            ]\n",
    "        }\n",
    "    }\n",
    "    return seqspec\n",
    "def generate_seqspecs(input_csv, output_dir):\n",
    "    os.makedirs(output_dir, exist_ok=True)\n",
    "    samples = {}\n",
    "    with open(input_csv, newline='') as csvfile:\n",
    "        reader = csv.DictReader(csvfile)\n",
    "        for row in reader:\n",
    "            prefix = os.path.basename(row[\"submitted_file_name\"]).split('_R')[0]\n",
    "            samples.setdefault(prefix, []).append(row)\n",
    "    for sample, reads in samples.items():\n",
    "        seqspec = build_seqspec(sample, reads)\n",
    "        output_file = os.path.join(output_dir, f\"seqspec_{sample}.yaml.gz\")\n",
    "        with gzip.open(output_file, 'wt', encoding='utf-8') as gzfile:\n",
    "           yaml.dump(seqspec,gzfile,sort_keys=False)\n",
    "        \n",
    "        print(f\"Generated: {output_file}\")\n",
    "    print(f\"\\nDone! Seqspecs saved in: {output_dir}\")\n",
    "def main():\n",
    "    parser = argparse.ArgumentParser(description=\"Generate seqspec YAMLs (gzipped) from CSV\")\n",
    "    parser.add_argument(\"-i\", \"--input\", required=True, help=\"Input CSV file\")\n",
    "    parser.add_argument(\"-o\", \"--output\", default=\"seqspec_files\", help=\"Output directory\")\n",
    "    args = parser.parse_args()\n",
    "    generate_seqspecs(args.input, args.output)\n",
    "if __name__ == \"__main__\":\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting /work/rr151/TF_perturbSeq50genes_BL/generate_seqspec10x5pv3-sgRNA.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile /work/rr151/TF_perturbSeq50genes_BL/generate_seqspec10x5pv3-sgRNA.py\n",
    "#!/usr/bin/env python3\n",
    "import csv\n",
    "import yaml\n",
    "import gzip\n",
    "import os\n",
    "import argparse\n",
    "\n",
    "# Custom representers so PyYAML outputs tags like !Assay\n",
    "def tag_representer(dumper, data):\n",
    "    # data should be a dict with exactly one key: \"!Tag\"\n",
    "    tag = list(data.keys())[0]\n",
    "    value = data[tag]\n",
    "    return dumper.represent_mapping(tag, value)\n",
    "# Register for the tags you use\n",
    "yaml.add_representer(dict, tag_representer)\n",
    "\n",
    "def build_seqspec(sample_name, reads):\n",
    "    \"\"\"\n",
    "    Build seqspec YAML for one sample, given its paired read files.\n",
    "    \"\"\"\n",
    "    reads.sort(key=lambda r: r['illumina_read_type'])  # Ensure R1 then R2\n",
    "    # Build !Read entries\n",
    "    read_entries = []\n",
    "    for r in reads:\n",
    "        read_tag = \"!Read\"\n",
    "        read_entry = {\n",
    "            read_tag:  {\n",
    "                \"read_id\": r[\"accession\"],\n",
    "                \"name\": r[\"read_names\"].strip(\"[]\\\"\"),\n",
    "                \"modality\": \"crispr\",\n",
    "                \"primer_id\": \"truseq_read1\" if r[\"illumina_read_type\"] == \"R1\" else \"truseq_read2\",\n",
    "                \"min_len\": 26 if r[\"illumina_read_type\"] == \"R1\" else 90,\n",
    "                \"max_len\": 26 if r[\"illumina_read_type\"] == \"R1\" else 90,\n",
    "                \"strand\": \"pos\" if r[\"illumina_read_type\"] == \"R1\" else \"neg\",\n",
    "                \"files\": [\n",
    "                    {\"!File\": {\n",
    "                        \"file_id\": r[\"accession\"],\n",
    "                        \"filename\": f\"https://api.data.igvf.org/sequence-files/{r['accession']}/@@download/{r['accession']}.fastq.gz\",\n",
    "                        \"filesize\": int(r[\"file_size\"]),\n",
    "                        \"filetype\": r[\"file_format\"],\n",
    "                        \"url\": f\"https://api.data.igvf.org/sequence-files/{r['accession']}/@@download/{r['accession']}.fastq.gz\",\n",
    "                        \"urltype\": \"https\",\n",
    "                        \"md5\": r[\"md5sum\"]\n",
    "                    }}\n",
    "                ]\n",
    "            }\n",
    "        \n",
    "        }\n",
    "        read_entries.append(read_entry)\n",
    "    index5 = reads[0]['index5']\n",
    "    index7 = reads[0]['index7']\n",
    "    seqspec = {\n",
    "        \"!Assay\": {\n",
    "            \"seqspec_version\": \"0.3.0\",\n",
    "            \"assay_id\": \"10XCRISPR5prime\",\n",
    "            \"name\": \"10x CRISPR and RNA assay 5'\",\n",
    "            \"doi\": \"https://doi.org/\",\n",
    "            \"date\": \"25 July 2025\",\n",
    "            \"description\": \"10x single-cell RNA-seq and CRISPR 5'\",\n",
    "            \"modalities\": [\"crispr\"],\n",
    "            \"lib_struct\": \"https://teichlab.github.io/scg_lib_structs/methods_html/10xChromium5vdjfb.html\",\n",
    "            \"sequence_protocol\": \"Illumina NextSeq 2000 (EFO:0010963), Illumina NovaSeq X Plus (EFO:0022841)\",\n",
    "            \"sequence_kit\": reads[0][\"sequencing_kit\"],\n",
    "            \"library_protocol\": [\n",
    "                {\"!LibProtocol\": {\n",
    "                    \"protocol_id\": \"Chromium Next GEM Single Cell 5' HT Reagent Kits v2 (Dual Index) (10x Genomics)\",\n",
    "                    \"name\": \"5' Gene Expression (GEX) Library Construction\",\n",
    "                    \"modality\": \"crispr\"\n",
    "                }}\n",
    "            ],\n",
    "            \"library_kit\": \"10XCRISPR5prime\",\n",
    "            \"sequence_spec\": read_entries,\n",
    "            \"library_spec\": [\n",
    "                {\"!Region\": {\n",
    "                    \"parent_id\": \"crispr\",\n",
    "                    \"region_id\": \"crispr\",\n",
    "                    \"region_type\": \"crispr\",\n",
    "                    \"name\": \"crispr\",\n",
    "                    \"sequence_type\": \"joined\",\n",
    "                    \"sequence\": \"AATGATACGGCGACCACCGAGATCTACACNNNNNNNNNNACACTCTTTCCCTACACGACGCTCTTCCGATCTNNNNNNNNNNNNNNNNXXXXXXXXXXXXTTTCTTATATGGGGNNNNNNNNNNNNNNNNNNNNGTTTAAGAGCTAAGCTGGAAACAGCATAGCAAGTTTAAATAAGGCTAGTCCGTTATCAACTTGAGATCGGAAGAGCACACGTCTGAACTCCAGTCACNNNNNNNNNNATCTCGTATGCCGTCTTCTGCTTG\",\n",
    "                    \"min_len\": 135,\n",
    "                    \"max_len\": 269,\n",
    "                    \"onlist\": None,\n",
    "                    \"regions\": [\n",
    "                        {\"!Region\": {\"region_id\": \"truseq_read1\", \"region_type\": \"truseq_read1\", \"name\": \"truseq_read1\", \"sequence_type\": \"fixed\", \"sequence\": \"ACACTCTTTCCCTACACGACGCTCTTCCGATCT\", \"min_len\": 33, \"max_len\": 33, \"onlist\": None, \"regions\": None, \"parent_id\": \"crispr\"}},\n",
    "                        {\"!Region\": {\"region_id\": \"cell_bc\", \"region_type\": \"barcode\", \"name\": \"cell_bc\", \"sequence_type\": \"onlist\", \"sequence\": \"NNNNNNNNNNNNNNNN\", \"min_len\": 16, \"max_len\": 16, \"onlist\": {\"!Onlist\": {\"file_id\": \"IGVFFI1697XAXI\", \"filename\": \"https://api.data.igvf.org/tabular-files/IGVFFI1697XAXI/@@download/IGVFFI1697XAXI.tsv.gz\", \"filetype\": \"tsv\", \"filesize\": 6900000, \"url\": \"https://api.data.igvf.org/tabular-files/IGVFFI1697XAXI/@@download/IGVFFI1697XAXI.tsv.gz\", \"urltype\": \"https\", \"md5\": \"5033d08c9b8d5080d2e5f65a3896b53b\"}}, \"regions\": None, \"parent_id\": \"crispr\"}},\n",
    "                        {\"!Region\": {\"region_id\": \"umi\", \"region_type\": \"umi\", \"name\": \"umi\", \"sequence_type\": \"random\", \"sequence\": \"XXXXXXXXXXXX\", \"min_len\": 12, \"max_len\": 12, \"onlist\": None, \"regions\": None, \"parent_id\": \"crispr\"}},\n",
    "                        {\"!Region\": {\"region_id\": \"tso\", \"region_type\": \"linker\", \"name\": \"Template switch oligo\", \"sequence_type\": \"fixed\", \"sequence\": \"TTTCTTATATGGG\", \"min_len\": 13, \"max_len\": 13, \"onlist\": None, \"regions\": None, \"parent_id\": \"crispr\"}},\n",
    "                        {\"!Region\": {\"region_id\": \"sgrna_target\", \"region_type\": \"sgrna_target\", \"name\": \"sgrna_target\", \"sequence_type\": \"onlist\", \"sequence\": \"NNNNNNNNNNNNNNNNNNNN\", \"min_len\": 20, \"max_len\": 20, \"onlist\": {\"!Onlist\": {\"file_id\": \"IGVFFI6638COVL\", \"filename\": \"https://api.data.igvf.org/reference-files/IGVFFI6638COVL/@@download/IGVFFI6638COVL.tsv.gz\", \"filetype\": \"tsv\", \"filesize\": 2800, \"url\": \"https://api.data.igvf.org/reference-files/IGVFFI6638COVL/@@download/IGVFFI6638COVL.tsv.gz\", \"urltype\": \"https\", \"md5\": \"7e6d956b7bf03c4cc6d16ae5fed9661b\"}}, \"regions\": None, \"parent_id\": \"crispr\"}},\n",
    "                        {\"!Region\": {\"region_id\": \"sgrna_scaffold\", \"region_type\": \"linker\", \"name\": \"sgrna_scaffold\", \"sequence_type\": \"fixed\", \"sequence\": \"GTTTAAGAGCTAAGCTGGAAACAGCATAGCAAGTTTA\", \"min_len\": 37, \"max_len\": 37, \"onlist\": None, \"regions\": None, \"parent_id\": \"crispr\"}},\n",
    "                        {\"!Region\": {\"region_id\": \"feature_primer\", \"region_type\": \"custom_primer\", \"name\": \"feature_primer\", \"sequence_type\": \"fixed\", \"sequence\": \"AATAAGGCTAGTCCGTTATCAACTTG\", \"min_len\": 26, \"max_len\": 26, \"onlist\": None, \"regions\": None, \"parent_id\": \"crispr\"}},\n",
    "                        {\"!Region\": {\"region_id\": \"truseq_read2\", \"region_type\": \"truseq_read2\", \"name\": \"truseq_read2\", \"sequence_type\": \"fixed\", \"sequence\": \"AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC\", \"min_len\": 34, \"max_len\": 34, \"onlist\": None, \"regions\": None, \"parent_id\": \"crispr\"}},\n",
    "                        {\"!Region\": {\"region_id\": \"illumina_p5\", \"region_type\": \"illumina_p5\", \"name\": \"Illumina P5\", \"sequence_type\": \"fixed\", \"sequence\": \"AATGATACGGCGACCACCGAGATCTACAC\", \"min_len\": 29, \"max_len\": 29, \"onlist\": None, \"regions\": None, \"parent_id\": \"crispr\"}},\n",
    "                        {\"!Region\": {\"region_id\": \"illumina_p7\", \"region_type\": \"illumina_p7\", \"name\": \"Illumina P7\", \"sequence_type\": \"fixed\", \"sequence\": \"ATCTCGTATGCCGTCTTCTGCTTG\", \"min_len\": 24, \"max_len\": 24, \"onlist\": None, \"regions\": None, \"parent_id\": \"crispr\"}},\n",
    "                        {\"!Region\": {\"region_id\": \"index5\", \"region_type\": \"index5\", \"name\": \"index5\", \"sequence_type\": \"fixed\", \"sequence\": index5, \"min_len\": 10, \"max_len\": 10, \"onlist\": None, \"regions\": None, \"parent_id\": \"crispr\"}},\n",
    "                        {\"!Region\": {\"region_id\": \"index7\", \"region_type\": \"index7\", \"name\": \"index7\", \"sequence_type\": \"fixed\", \"sequence\": index7, \"min_len\": 10, \"max_len\": 10, \"onlist\": None, \"regions\": None, \"parent_id\": \"crispr\"}}\n",
    "                    ]\n",
    "                }}\n",
    "            ]\n",
    "        }\n",
    "    }\n",
    "    return seqspec\n",
    "def generate_seqspecs(input_csv, output_dir):\n",
    "    os.makedirs(output_dir, exist_ok=True)\n",
    "    samples = {}\n",
    "    with open(input_csv, newline='') as csvfile:\n",
    "        reader = csv.DictReader(csvfile)\n",
    "        for row in reader:\n",
    "            prefix = os.path.basename(row[\"submitted_file_name\"]).split('_R')[0]\n",
    "            samples.setdefault(prefix, []).append(row)\n",
    "    for sample, reads in samples.items():\n",
    "        seqspec = build_seqspec(sample, reads)\n",
    "        output_file = os.path.join(output_dir, f\"seqspec_{sample}.yaml.gz\")\n",
    "        with gzip.open(output_file, 'wt', encoding='utf-8') as gzfile:\n",
    "           yaml.dump(seqspec,gzfile,sort_keys=False)\n",
    "        \n",
    "        print(f\"Generated: {output_file}\")\n",
    "    print(f\"\\nDone! Seqspecs saved in: {output_dir}\")\n",
    "def main():\n",
    "    parser = argparse.ArgumentParser(description=\"Generate seqspec YAMLs (gzipped) from CSV\")\n",
    "    parser.add_argument(\"-i\", \"--input\", required=True, help=\"Input CSV file\")\n",
    "    parser.add_argument(\"-o\", \"--output\", default=\"seqspec_files\", help=\"Output directory\")\n",
    "    args = parser.parse_args()\n",
    "    generate_seqspecs(args.input, args.output)\n",
    "if __name__ == \"__main__\":\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "python3 /work/rr151/TF_perturbSeq50genes_BL/generate_seqspec10x5pv2-sgRNA.py -i /work/rr151/TF_perturbSeq50genes_BL/10x5pv2-sgRNA.csv -o /work/rr151/TF_perturbSeq50genes_BL/seqspecCORRECTED9112025\n",
    "python3 /work/rr151/TF_perturbSeq50genes_BL/generate_seqspec10x5pv2_gex.py -i /work/rr151/TF_perturbSeq50genes_BL/10x5pv2-GEX.csv -o /work/rr151/TF_perturbSeq50genes_BL/seqspec_yaml\n",
    "python3 /work/rr151/TF_perturbSeq50genes_BL/generate_seqspec10x5pv3_gex.py -i /work/rr151/TF_perturbSeq50genes_BL/10x5pv3-GEX.csv -o /work/rr151/TF_perturbSeq50genes_BL/seqspecCORRECTED9102025\n",
    "python3 /work/rr151/TF_perturbSeq50genes_BL/generate_seqspec10x5pv3-sgRNA.py -i /work/rr151/TF_perturbSeq50genes_BL/10x5pv3-sgRNA.csv -o /work/rr151/TF_perturbSeq50genes_BL/seqspecCORRECTED9112025"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
